src/Npgsql/NpgsqlCommand.cs

```diff
@@ -1401,12 +1402,25 @@ void LogCommand()
                 var p = s.InputParameters;
                 if (p.Count > 0 && (NpgsqlLogManager.IsParameterLoggingEnabled || connector.Settings.LogParameters))
                 {
-                    sb.Append('\t').Append("Parameters:");
                     for (var i = 0; i < p.Count; i++)
-                        sb.Append("\t$").Append(i + 1).Append(": ").Append(Convert.ToString(p[i].Value, CultureInfo.InvariantCulture));
+                    {
+                        if (p[i].Value == null || p[i].Value == DBNull.Value)
+                            sb.Append("\t").Append("Parameters $").Append(i + 1).Append(":\t").Append(Convert.ToString("null", CultureInfo.InvariantCulture));
+                        else if (p[i].Value is IList list)
+                        {
+                            foreach (var val in list)
+                            {
+                                if (list.IndexOf(val) == 0)
```

Comment:
This looks like it needlessly rescans the list for every value. Seem like you can extract the common part (`sb.Append("\t").Append("Parameters $").Append(i + 1)`) outside the conditions, and use a for loop and check the index to know if it's the first one.

Alternatively, C# 8 pattern matching could probably make this much prettier, but also less efficient (we'd be constructing strings instead of only calling StringBuilder.Append).

Reply:
Thanks @Roji, fixed this and have implemted the c# 8 pattern matching and replaced the IndexOf.